--- wxsqlite3.old	2011-03-07 23:34:44.000000000 +0300
+++ wxsqlite3.cpp	2011-03-31 14:45:55.000000000 +0400
@@ -30,6 +30,24 @@
 #include "wx/wxsqlite3.h"
 #include "wx/wxsqlite3opt.h"
 
+// Error handling in wxSQLite3 library
+
+#if WXSQLITE_USE_EXCEPTIONS
+    #define WXSQLITE3_ERROR(code, message, result) throw wxSQLite3Exception(code, message);
+#else
+    void wxSQLite3_Error(int errorCode, const wxString& errorMsg)
+    {
+        if (errorCode == 9) return;
+        wxString m_errorMessage =
+            wxSQLite3Exception::ErrorCodeAsString(errorCode) + wxT("[") +
+            wxString::Format(wxT("%d"), errorCode) + wxT("]: ") +
+            wxGetTranslation(errorMsg);
+
+        wxLogError(m_errorMessage);
+    }
+    #define WXSQLITE3_ERROR(code, message, result) wxSQLite3_Error(code, message); return result;
+#endif  // WXSQLITE_USE_EXCEPTIONS
+
 // Suppress some Visual C++ warnings regarding the default constructor
 // for a C struct used only in SQLite modules
 #ifdef __VISUALC__
@@ -72,7 +89,7 @@
   if (! s_dll.Load(wxT("libsqlite3")))
 #endif
   {
-    throw wxSQLite3Exception(-1, wxT("error loading dynamic library"));
+    WXSQLITE3_ERROR(-1, wxT("error loading dynamic library"), );
   }
 
 #define DYNFUNC(rcode, rtype, fname, farg, farguse) \
@@ -80,7 +97,7 @@
   if (! s_##fname)\
   {\
     s_dll.Unload();\
-    throw wxSQLite3Exception(-1, wxT("error getting symbol <") wxT(#fname) wxT(">"));\
+    WXSQLITE3_ERROR(-1, wxT("error getting symbol <") wxT(#fname) wxT(">"), );\
   }
 #include "wx/wxsqlite3dyn.h"
 #undef DYNFUNC
@@ -439,7 +456,7 @@
 
   if (columnIndex < 0 || columnIndex > m_cols-1)
   {
-    throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_INVALID_INDEX);
+    WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_INVALID_INDEX, wxEmptyString);
   }
 
   const char* localValue = (const char*) sqlite3_column_text((sqlite3_stmt*) m_stmt, columnIndex);
@@ -533,7 +550,7 @@
 
   if (columnIndex < 0 || columnIndex > m_cols-1)
   {
-    throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_INVALID_INDEX);
+    WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_INVALID_INDEX, NULL);
   }
 
   len = sqlite3_column_bytes((sqlite3_stmt*) m_stmt, columnIndex);
@@ -552,7 +569,7 @@
 
   if (columnIndex < 0 || columnIndex > m_cols-1)
   {
-    throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_INVALID_INDEX);
+    WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_INVALID_INDEX, buffer);
   }
 
   int len = sqlite3_column_bytes((sqlite3_stmt*) m_stmt, columnIndex);
@@ -757,7 +774,7 @@
     }
   }
 
-  throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_INVALID_INDEX);
+  WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_INVALID_INDEX, -1);
 }
 
 wxString wxSQLite3ResultSet::GetColumnName(int columnIndex)
@@ -766,7 +783,7 @@
 
   if (columnIndex < 0 || columnIndex > m_cols-1)
   {
-    throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_INVALID_INDEX);
+    WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_INVALID_INDEX, wxEmptyString);
   }
 
   const char* localValue = sqlite3_column_name((sqlite3_stmt*) m_stmt, columnIndex);
@@ -779,7 +796,7 @@
 
   if (columnIndex < 0 || columnIndex > m_cols-1)
   {
-    throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_INVALID_INDEX);
+    WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_INVALID_INDEX, wxEmptyString);
   }
 
   const char* localValue = sqlite3_column_decltype((sqlite3_stmt*) m_stmt, columnIndex);
@@ -792,7 +809,7 @@
 
   if (columnIndex < 0 || columnIndex > m_cols-1)
   {
-    throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_INVALID_INDEX);
+    WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_INVALID_INDEX, -1);
   }
 
   return sqlite3_column_type((sqlite3_stmt*) m_stmt, columnIndex);
@@ -833,7 +850,7 @@
     rc = sqlite3_finalize((sqlite3_stmt*) m_stmt);
     m_stmt = 0;
     const char* localError = sqlite3_errmsg((sqlite3*) m_db);
-    throw wxSQLite3Exception(rc, wxString::FromUTF8(localError));
+    WXSQLITE3_ERROR(rc, wxString::FromUTF8(localError), false);
   }
 }
 
@@ -846,7 +863,7 @@
     if (rc != SQLITE_OK)
     {
       const char* localError = sqlite3_errmsg((sqlite3*) m_db);
-      throw wxSQLite3Exception(rc, wxString::FromUTF8(localError));
+      WXSQLITE3_ERROR(rc, wxString::FromUTF8(localError), );
     }
   }
 }
@@ -871,7 +888,7 @@
 {
   if (m_stmt == 0)
   {
-    throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_NOSTMT);
+    return;
   }
 }
 
@@ -881,7 +898,7 @@
   CheckStmt();
   if (columnIndex < 0 || columnIndex > m_cols-1)
   {
-    throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_INVALID_INDEX);
+    WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_INVALID_INDEX, wxEmptyString);
   }
 
   const char* localValue = sqlite3_column_database_name((sqlite3_stmt*) m_stmt, columnIndex);
@@ -891,7 +908,7 @@
     return wxEmptyString;
 #else
   wxUnusedVar(columnIndex);
-  throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_NOMETADATA);
+  WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_NOMETADATA, wxEmptyString);
 #endif
 }
 
@@ -901,7 +918,7 @@
   CheckStmt();
   if (columnIndex < 0 || columnIndex > m_cols-1)
   {
-    throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_INVALID_INDEX);
+    WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_INVALID_INDEX, wxEmptyString);
   }
 
   const char* localValue = sqlite3_column_table_name((sqlite3_stmt*) m_stmt, columnIndex);
@@ -911,7 +928,7 @@
     return wxEmptyString;
 #else
   wxUnusedVar(columnIndex);
-  throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_NOMETADATA);
+  WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_NOMETADATA, wxEmptyString);
 #endif
 }
 
@@ -921,7 +938,7 @@
   CheckStmt();
   if (columnIndex < 0 || columnIndex > m_cols-1)
   {
-    throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_INVALID_INDEX);
+    WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_INVALID_INDEX, wxEmptyString);
   }
 
   const char* localValue = sqlite3_column_origin_name((sqlite3_stmt*) m_stmt, columnIndex);
@@ -931,7 +948,7 @@
     return wxEmptyString;
 #else
   wxUnusedVar(columnIndex);
-  throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_NOMETADATA);
+  WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_NOMETADATA, wxEmptyString);
 #endif
 }
 
@@ -1036,14 +1053,14 @@
     }
   }
 
-  throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_INVALID_NAME);
+  WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_INVALID_NAME, -1);
 }
 
 wxString wxSQLite3Table::GetAsString(int columnIndex)
 {
   if (columnIndex < 0 || columnIndex > m_cols-1)
   {
-    throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_INVALID_INDEX);
+    WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_INVALID_INDEX, wxEmptyString);
   }
 
   int nIndex = (m_currentRow*m_cols) + m_cols + columnIndex;
@@ -1212,7 +1229,7 @@
   {
     if (columnIndex < 0 || columnIndex > m_cols-1)
     {
-      throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_INVALID_INDEX);
+      WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_INVALID_INDEX, nullValue);
     }
     int nIndex = (m_currentRow*m_cols) + m_cols + columnIndex;
     return wxSQLite3AtoF(m_results[nIndex]);
@@ -1326,7 +1343,7 @@
 
   if (columnIndex < 0 || columnIndex > m_cols-1)
   {
-    throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_INVALID_INDEX);
+    WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_INVALID_INDEX, false);
   }
 
   int index = (m_currentRow*m_cols) + m_cols + columnIndex;
@@ -1346,7 +1363,7 @@
 
   if (columnIndex < 0 || columnIndex > m_cols-1)
   {
-    throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_INVALID_INDEX);
+    WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_INVALID_INDEX, wxEmptyString);
   }
 
   const char* localValue = m_results[columnIndex];
@@ -1359,7 +1376,7 @@
 
   if (row < 0 || row > m_rows-1)
   {
-    throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_INVALID_ROW);
+    WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_INVALID_ROW, );
   }
 
   m_currentRow = row;
@@ -1374,7 +1391,7 @@
 {
   if (m_results == 0)
   {
-    throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_NORESULT);
+    WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_NORESULT, );
   }
 }
 
@@ -1454,7 +1471,7 @@
     if (rc != SQLITE_OK)
     {
       localError = sqlite3_errmsg((sqlite3*) m_db);
-      throw wxSQLite3Exception(rc, wxString::FromUTF8(localError));
+      WXSQLITE3_ERROR(rc, wxString::FromUTF8(localError), -1);
     }
 
     return rowsChanged;
@@ -1463,7 +1480,7 @@
   {
     rc = sqlite3_reset((sqlite3_stmt*) m_stmt);
     localError = sqlite3_errmsg((sqlite3*) m_db);
-    throw wxSQLite3Exception(rc, wxString::FromUTF8(localError));
+    WXSQLITE3_ERROR(rc, wxString::FromUTF8(localError), -1);
   }
 }
 
@@ -1479,7 +1496,7 @@
     }
     else
     {
-      throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_NOTOWNED);
+      WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_NOTOWNED, wxSQLite3ResultSet());
     }
   }
 
@@ -1497,7 +1514,7 @@
   {
     rc = sqlite3_reset((sqlite3_stmt*) m_stmt);
     const char* localError = sqlite3_errmsg((sqlite3*) m_db);
-    throw wxSQLite3Exception(rc, wxString::FromUTF8(localError));
+    WXSQLITE3_ERROR(rc, wxString::FromUTF8(localError), wxSQLite3ResultSet());
   }
 }
 
@@ -1535,7 +1552,7 @@
 
   if (rc != SQLITE_OK)
   {
-    throw wxSQLite3Exception(rc, wxERRMSG_BIND_STR);
+    WXSQLITE3_ERROR(rc, wxERRMSG_BIND_STR, );
   }
 }
 
@@ -1546,7 +1563,7 @@
 
   if (rc != SQLITE_OK)
   {
-    throw wxSQLite3Exception(rc, wxERRMSG_BIND_INT);
+    WXSQLITE3_ERROR(rc, wxERRMSG_BIND_INT, );
   }
 }
 
@@ -1557,7 +1574,7 @@
 
   if (rc != SQLITE_OK)
   {
-    throw wxSQLite3Exception(rc, wxERRMSG_BIND_INT64);
+    WXSQLITE3_ERROR(rc, wxERRMSG_BIND_INT64, );
   }
 }
 
@@ -1568,7 +1585,7 @@
 
   if (rc != SQLITE_OK)
   {
-    throw wxSQLite3Exception(rc, wxERRMSG_BIND_DBL);
+    WXSQLITE3_ERROR(rc, wxERRMSG_BIND_DBL, );
   }
 }
 
@@ -1579,7 +1596,7 @@
 
   if (rc != SQLITE_OK)
   {
-    throw wxSQLite3Exception(rc, wxERRMSG_BIND_STR);
+    WXSQLITE3_ERROR(rc, wxERRMSG_BIND_STR, );
   }
 }
 
@@ -1591,7 +1608,7 @@
 
   if (rc != SQLITE_OK)
   {
-    throw wxSQLite3Exception(rc, wxERRMSG_BIND_BLOB);
+    WXSQLITE3_ERROR(rc, wxERRMSG_BIND_BLOB, );
   }
 }
 
@@ -1604,7 +1621,7 @@
 
   if (rc != SQLITE_OK)
   {
-    throw wxSQLite3Exception(rc, wxERRMSG_BIND_BLOB);
+    WXSQLITE3_ERROR(rc, wxERRMSG_BIND_BLOB, );
   }
 }
 
@@ -1616,7 +1633,7 @@
   }
   else
   {
-    throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_BIND_DATETIME);
+    WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_BIND_DATETIME, );
   }
 }
 
@@ -1628,7 +1645,7 @@
   }
   else
   {
-    throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_BIND_DATETIME);
+    WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_BIND_DATETIME, );
   }
 }
 
@@ -1640,7 +1657,7 @@
   }
   else
   {
-    throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_BIND_DATETIME);
+    WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_BIND_DATETIME, );
   }
 }
 
@@ -1652,7 +1669,7 @@
   }
   else
   {
-    throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_BIND_DATETIME);
+    WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_BIND_DATETIME, );
   }
 }
 
@@ -1664,7 +1681,7 @@
   }
   else
   {
-    throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_BIND_DATETIME);
+    WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_BIND_DATETIME, );
   }
 }
 
@@ -1676,7 +1693,7 @@
   }
   else
   {
-    throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_BIND_DATETIME);
+    WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_BIND_DATETIME, );
   }
 }
 
@@ -1692,7 +1709,7 @@
 
   if (rc != SQLITE_OK)
   {
-    throw wxSQLite3Exception(rc, wxERRMSG_BIND_NULL);
+    WXSQLITE3_ERROR(rc, wxERRMSG_BIND_NULL, );
   }
 }
 
@@ -1703,12 +1720,12 @@
   int rc = sqlite3_bind_zeroblob((sqlite3_stmt*) m_stmt, paramIndex, blobSize);
   if (rc != SQLITE_OK)
   {
-    throw wxSQLite3Exception(rc, wxERRMSG_BIND_ZEROBLOB);
+    WXSQLITE3_ERROR(rc, wxERRMSG_BIND_ZEROBLOB, );
   }
 #else
   wxUnusedVar(paramIndex);
   wxUnusedVar(blobSize);
-  throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_NOINCBLOB);
+  WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_NOINCBLOB, );
 #endif
 }
 
@@ -1720,7 +1737,7 @@
 
   if (rc != SQLITE_OK)
   {
-    throw wxSQLite3Exception(rc, wxERRMSG_BIND_CLEAR);
+    WXSQLITE3_ERROR(rc, wxERRMSG_BIND_CLEAR, );
   }
 #else
   for (int paramIndex = 1; paramIndex <= GetParamCount(); paramIndex++)
@@ -1750,7 +1767,7 @@
     if (rc != SQLITE_OK)
     {
       const char* localError = sqlite3_errmsg((sqlite3*) m_db);
-      throw wxSQLite3Exception(rc, wxString::FromUTF8(localError));
+      WXSQLITE3_ERROR(rc, wxString::FromUTF8(localError), );
     }
   }
 }
@@ -1776,7 +1793,7 @@
     if (rc != SQLITE_OK)
     {
       const char* localError = sqlite3_errmsg((sqlite3*) m_db);
-      throw wxSQLite3Exception(rc, wxString::FromUTF8(localError));
+      WXSQLITE3_ERROR(rc, wxString::FromUTF8(localError), );
     }
   }
 }
@@ -1790,7 +1807,7 @@
 {
   if (m_db == 0)
   {
-    throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_NODB);
+    WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_NODB, );
   }
 }
 
@@ -1798,7 +1815,7 @@
 {
   if (m_stmt == 0)
   {
-    throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_NOSTMT);
+    return;
   }
 }
 
@@ -1870,7 +1887,7 @@
   if (rc != SQLITE_OK)
   {
     const char* localError = sqlite3_errmsg((sqlite3*) m_db);
-    throw wxSQLite3Exception(rc, wxString::FromUTF8(localError));
+    WXSQLITE3_ERROR(rc, wxString::FromUTF8(localError), blobValue);
   }
 
   blobValue.UngetAppendBuf((size_t) length);
@@ -1878,7 +1895,7 @@
   wxUnusedVar(blobValue);
   wxUnusedVar(length);
   wxUnusedVar(offset);
-  throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_NOINCBLOB);
+  WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_NOINCBLOB, blobValue);
 #endif
   return blobValue;
 }
@@ -1896,17 +1913,17 @@
     if (rc != SQLITE_OK)
     {
       const char* localError = sqlite3_errmsg((sqlite3*) m_db);
-      throw wxSQLite3Exception(rc, wxString::FromUTF8(localError));
+      WXSQLITE3_ERROR(rc, wxString::FromUTF8(localError), );
     }
   }
   else
   {
-    throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_INVALID_BLOB);
+    WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_INVALID_BLOB, );
   }
 #else
   wxUnusedVar(blobValue);
   wxUnusedVar(offset);
-  throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_NOINCBLOB);
+  WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_NOINCBLOB, );
 #endif
 }
 
@@ -1926,7 +1943,7 @@
   CheckBlob();
   return sqlite3_blob_bytes((sqlite3_blob*) m_blob);
 #else
-  throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_NOINCBLOB);
+  WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_NOINCBLOB, 0);
   return 0;
 #endif
 }
@@ -1939,11 +1956,11 @@
   if (rc != SQLITE_OK)
   {
     const char* localError = sqlite3_errmsg((sqlite3*) m_db);
-    throw wxSQLite3Exception(rc, wxString::FromUTF8(localError));
+    WXSQLITE3_ERROR(rc, wxString::FromUTF8(localError), );
   }
 #else
   wxUnusedVar(rowid);
-  throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_NOBLOBREBIND);
+  WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_NOBLOBREBIND, );
 #endif
 }
 
@@ -1958,11 +1975,11 @@
     if (rc != SQLITE_OK)
     {
       const char* localError = sqlite3_errmsg((sqlite3*) m_db);
-      throw wxSQLite3Exception(rc, wxString::FromUTF8(localError));
+      WXSQLITE3_ERROR(rc, wxString::FromUTF8(localError), );
     }
   }
 #else
-  throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_NOINCBLOB);
+  WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_NOINCBLOB, );
 #endif
 }
 
@@ -1970,7 +1987,7 @@
 {
   if (!m_ok)
   {
-    throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_INVALID_BLOB);
+    WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_INVALID_BLOB, );
   }
 }
 
@@ -1987,7 +2004,7 @@
   int rc = sqlite3_enable_shared_cache(flag);
   if (rc != SQLITE_OK)
   {
-    throw wxSQLite3Exception(rc, wxERRMSG_SHARED_CACHE);
+    WXSQLITE3_ERROR(rc, wxERRMSG_SHARED_CACHE, );
   }
   ms_sharedCacheEnabled = enable;
 }
@@ -2119,7 +2136,7 @@
     }
     else
     {
-      throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_DBASSIGN_FAILED);
+      WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_DBASSIGN_FAILED, *this);
     }
   }
   return *this;
@@ -2148,7 +2165,7 @@
   {
     const char* localError = sqlite3_errmsg((sqlite3*) m_db);
     Close();
-    throw wxSQLite3Exception(rc, wxString::FromUTF8(localError));
+    WXSQLITE3_ERROR(rc, wxString::FromUTF8(localError), );
   }
 
   rc = sqlite3_extended_result_codes((sqlite3*) m_db, 1);
@@ -2156,7 +2173,7 @@
   {
     const char* localError = sqlite3_errmsg((sqlite3*) m_db);
     Close();
-    throw wxSQLite3Exception(rc, wxString::FromUTF8(localError));
+    WXSQLITE3_ERROR(rc, wxString::FromUTF8(localError), );
   }
 
 #if WXSQLITE3_HAVE_CODEC
@@ -2167,7 +2184,7 @@
     {
       const char* localError = sqlite3_errmsg((sqlite3*) m_db);
       Close();
-      throw wxSQLite3Exception(rc, wxString::FromUTF8(localError));
+      WXSQLITE3_ERROR(rc, wxString::FromUTF8(localError), );
     }
     m_isEncrypted = true;
   }
@@ -2234,7 +2251,7 @@
   if (rc != SQLITE_OK)
   {
     sqlite3_close(pDest);
-    throw wxSQLite3Exception(rc, wxERRMSG_DBOPEN_FAILED);
+    WXSQLITE3_ERROR(rc, wxERRMSG_DBOPEN_FAILED, );
   }
 #if WXSQLITE3_HAVE_CODEC
   if (key.GetDataLen() > 0)
@@ -2244,7 +2261,7 @@
     {
       const char* localError = sqlite3_errmsg((sqlite3*) pDest);
       sqlite3_close(pDest);
-      throw wxSQLite3Exception(rc, wxString::FromUTF8(localError));
+      WXSQLITE3_ERROR(rc, wxString::FromUTF8(localError), );
     }
   }
 #else
@@ -2256,7 +2273,7 @@
   {
     const char* localError = sqlite3_errmsg(pDest);
     sqlite3_close(pDest);
-    throw wxSQLite3Exception(rc, wxString::FromUTF8(localError));
+    WXSQLITE3_ERROR(rc, wxString::FromUTF8(localError), );
   }
 
   do
@@ -2282,12 +2299,12 @@
   {
     const char* localError = sqlite3_errmsg(pDest);
     sqlite3_close(pDest);
-    throw wxSQLite3Exception(rc, wxString::FromUTF8(localError));
+    WXSQLITE3_ERROR(rc, wxString::FromUTF8(localError), );
   }
 #else
   wxUnusedVar(targetFileName);
   wxUnusedVar(sourceDatabaseName);
-  throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_NOBACKUP);
+  WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_NOBACKUP, );
 #endif
 }
 
@@ -2322,7 +2339,7 @@
   if (rc != SQLITE_OK)
   {
     sqlite3_close(pSrc);
-    throw wxSQLite3Exception(rc, wxERRMSG_DBOPEN_FAILED);
+    WXSQLITE3_ERROR(rc, wxERRMSG_DBOPEN_FAILED, );
   }
 #if WXSQLITE3_HAVE_CODEC
   if (key.GetDataLen() > 0)
@@ -2332,7 +2349,7 @@
     {
       const char* localError = sqlite3_errmsg((sqlite3*) pSrc);
       sqlite3_close(pSrc);
-      throw wxSQLite3Exception(rc, wxString::FromUTF8(localError));
+      WXSQLITE3_ERROR(rc, wxString::FromUTF8(localError), );
     }
   }
 #else
@@ -2344,7 +2361,7 @@
   {
     const char* localError = sqlite3_errmsg((sqlite3*) m_db);
     sqlite3_close(pSrc);
-    throw wxSQLite3Exception(rc, wxString::FromUTF8(localError));
+    WXSQLITE3_ERROR(rc, wxString::FromUTF8(localError), );
   }
   while ((rc = sqlite3_backup_step(pBackup, 100)) == SQLITE_OK || rc == SQLITE_BUSY)
   {
@@ -2362,18 +2379,18 @@
   else if (rc == SQLITE_BUSY || rc == SQLITE_LOCKED)
   {
     sqlite3_close(pSrc);
-    throw wxSQLite3Exception(rc, wxERRMSG_SOURCEDB_BUSY);
+    WXSQLITE3_ERROR(rc, wxERRMSG_SOURCEDB_BUSY, );
   }
   else
   {
     const char* localError = sqlite3_errmsg(pSrc);
     sqlite3_close(pSrc);
-    throw wxSQLite3Exception(rc, wxString::FromUTF8(localError));
+    WXSQLITE3_ERROR(rc, wxString::FromUTF8(localError), );
   }
 #else
   wxUnusedVar(sourceFileName);
   wxUnusedVar(targetDatabaseName);
-  throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_NOBACKUP);
+  WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_NOBACKUP, );
 #endif
 }
 
@@ -2431,7 +2448,7 @@
   ExecuteUpdate(wxString(wxT("savepoint "))+savepointName);
 #else
   wxUnusedVar(savepointName);
-  throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_NOSAVEPOINT);
+  WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_NOSAVEPOINT, );
 #endif
 }
 
@@ -2441,7 +2458,7 @@
   ExecuteUpdate(wxString(wxT("release savepoint "))+savepointName);
 #else
   wxUnusedVar(savepointName);
-  throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_NOSAVEPOINT);
+  WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_NOSAVEPOINT, );
 #endif
 }
 
@@ -2663,7 +2680,7 @@
   }
   else
   {
-    throw wxSQLite3Exception(rc, wxString::FromUTF8(localError));
+    WXSQLITE3_ERROR(rc, wxString::FromUTF8(localError), 0);
   }
 }
 
@@ -2699,7 +2716,7 @@
   {
     rc = sqlite3_finalize(stmt);
     const char* localError= sqlite3_errmsg((sqlite3*) m_db);
-    throw wxSQLite3Exception(rc, wxString::FromUTF8(localError));
+    WXSQLITE3_ERROR(rc, wxString::FromUTF8(localError), wxSQLite3ResultSet());
   }
 }
 
@@ -2721,7 +2738,7 @@
 
   if (resultSet.Eof() || resultSet.GetColumnCount() < 1)
   {
-    throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_INVALID_QUERY);
+    WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_INVALID_QUERY, 0);
   }
 
   long value = 0;
@@ -2759,7 +2776,7 @@
   }
   else
   {
-    throw wxSQLite3Exception(rc, wxString::FromUTF8(localError));
+    WXSQLITE3_ERROR(rc, wxString::FromUTF8(localError), wxSQLite3Table());
   }
 }
 
@@ -2805,7 +2822,7 @@
   if (rc != SQLITE_OK)
   {
     const char* localError = sqlite3_errmsg((sqlite3*) m_db);
-    throw wxSQLite3Exception(rc, wxString::FromUTF8(localError));
+    WXSQLITE3_ERROR(rc, wxString::FromUTF8(localError), wxSQLite3Blob());
   }
   return wxSQLite3Blob(m_db, (void*) blobHandle, writable);
 #else
@@ -2814,7 +2831,7 @@
   wxUnusedVar(tableName);
   wxUnusedVar(dbName);
   wxUnusedVar(writable);
-  throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_NOINCBLOB);
+  WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_NOINCBLOB, wxSQLite3Blob());
   return NULL;
 #endif
 }
@@ -2957,7 +2974,7 @@
   }
 #else
   wxUnusedVar(walHook);
-  throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_NOWAL);
+  WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_NOWAL, );
 #endif
 }
 
@@ -2972,11 +2989,11 @@
   if (rc != SQLITE_OK)
   {
     const char* localError = sqlite3_errmsg((sqlite3*) m_db);
-    throw wxSQLite3Exception(rc, wxString::FromUTF8(localError));
+    WXSQLITE3_ERROR(rc, wxString::FromUTF8(localError), );
   }
 #else
   wxUnusedVar(database);
-  throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_NOWAL);
+  WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_NOWAL, );
 #endif
 }
 
@@ -2989,11 +3006,11 @@
   if (rc != SQLITE_OK)
   {
     const char* localError = sqlite3_errmsg((sqlite3*) m_db);
-    throw wxSQLite3Exception(rc, wxString::FromUTF8(localError));
+    WXSQLITE3_ERROR(rc, wxString::FromUTF8(localError), );
   }
 #else
   wxUnusedVar(frameCount);
-  throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_NOWAL);
+  WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_NOWAL, );
 #endif
 }
 
@@ -3020,7 +3037,7 @@
   if (rc != SQLITE_OK)
   {
     const char* localError = sqlite3_errmsg((sqlite3*) m_db);
-    throw wxSQLite3Exception(rc, wxString::FromUTF8(localError));
+    WXSQLITE3_ERROR(rc, wxString::FromUTF8(localError), );
   }
 }
 
@@ -3028,7 +3045,7 @@
 {
   if (!m_db)
   {
-    throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_NODB);
+    WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_NODB, );
   }
 }
 
@@ -3044,7 +3061,7 @@
   if (rc != SQLITE_OK)
   {
     const char* localError = sqlite3_errmsg((sqlite3*) m_db);
-    throw wxSQLite3Exception(rc, wxString::FromUTF8(localError));
+    WXSQLITE3_ERROR(rc, wxString::FromUTF8(localError), NULL);
   }
 
   return stmt;
@@ -3088,7 +3105,7 @@
   if (rc != SQLITE_OK)
   {
     const char* localError = sqlite3_errmsg((sqlite3*) m_db);
-    throw wxSQLite3Exception(rc, wxString::FromUTF8(localError));
+    WXSQLITE3_ERROR(rc, wxString::FromUTF8(localError), );
   }
 
   if (dataType      != NULL) *dataType      = wxString::FromUTF8(localDataType);
@@ -3106,7 +3123,7 @@
   wxUnusedVar(notNull);
   wxUnusedVar(primaryKey);
   wxUnusedVar(autoIncrement);
-  throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_NOMETADATA);
+  WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_NOMETADATA, );
 #endif
 }
 
@@ -3122,12 +3139,12 @@
   if (rc != SQLITE_OK)
   {
     const char* localError = sqlite3_errmsg((sqlite3*) m_db);
-    throw wxSQLite3Exception(rc, wxString::FromUTF8(localError));
+    WXSQLITE3_ERROR(rc, wxString::FromUTF8(localError), );
   }
 #else
   wxUnusedVar(fileName);
   wxUnusedVar(entryPoint);
-  throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_NOLOADEXT);
+  WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_NOLOADEXT, );
 #endif
 }
 
@@ -3139,11 +3156,11 @@
   if (rc != SQLITE_OK)
   {
     const char* localError = sqlite3_errmsg((sqlite3*) m_db);
-    throw wxSQLite3Exception(rc, wxString::FromUTF8(localError));
+    WXSQLITE3_ERROR(rc, wxString::FromUTF8(localError), );
   }
 #else
   wxUnusedVar(enable);
-  throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_NOLOADEXT);
+  WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_NOLOADEXT, );
 #endif
 }
 
@@ -3160,7 +3177,7 @@
   ReKey(binaryNewKey);
 #else
   wxUnusedVar(newKey);
-  throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_NOCODEC);
+  WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_NOCODEC, );
 #endif
 }
 
@@ -3171,11 +3188,11 @@
   if (rc != SQLITE_OK)
   {
     const char* localError = sqlite3_errmsg((sqlite3*) m_db);
-    throw wxSQLite3Exception(rc, wxString::FromUTF8(localError));
+    WXSQLITE3_ERROR(rc, wxString::FromUTF8(localError), );
   }
 #else
   wxUnusedVar(newKey);
-  throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_NOCODEC);
+  WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_NOCODEC, );
 #endif
 }
 
@@ -3238,7 +3255,7 @@
   int rc = sqlite3_initialize();
   if (rc != SQLITE_OK)
   {
-    throw wxSQLite3Exception(rc, wxERRMSG_INITIALIZE);
+    WXSQLITE3_ERROR(rc, wxERRMSG_INITIALIZE, );
   }
 #endif
 }
@@ -3250,7 +3267,7 @@
   int rc = sqlite3_shutdown();
   if (rc != SQLITE_OK)
   {
-    throw wxSQLite3Exception(rc, wxERRMSG_SHUTDOWN);
+    WXSQLITE3_ERROR(rc, wxERRMSG_SHUTDOWN, );
   }
 #endif
 }
@@ -4155,7 +4172,7 @@
   sqlite3_intarray* p = (sqlite3_intarray*) sqlite3_malloc( sizeof(*p) );
   if (p == 0)
   {
-    throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_NOMEM);
+    WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_NOMEM, wxSQLite3IntegerCollection());
   }
   p->n = 0;
   p->a= NULL;
@@ -4170,12 +4187,12 @@
   if (rc != SQLITE_OK)
   {
     const char* localError = sqlite3_errmsg((sqlite3*) m_db);
-    throw wxSQLite3Exception(rc, wxString::FromUTF8(localError));
+    WXSQLITE3_ERROR(rc, wxString::FromUTF8(localError), wxSQLite3IntegerCollection());
   }
   return wxSQLite3IntegerCollection(collectionName, p);
 #else
   wxUnusedVar(collectionName);
-  throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_NOCOLLECTIONS);
+  WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_NOCOLLECTIONS, );
 #endif // WXSQLITE3_USE_NAMED_COLLECTIONS
 }
 
@@ -4253,7 +4270,7 @@
   sqlite3_chararray* p = (sqlite3_chararray*) sqlite3_malloc( sizeof(*p) );
   if (p == 0)
   {
-    throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_NOMEM);
+    WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_NOMEM, wxSQLite3StringCollection());
   }
   p->n = 0;
   p->a= NULL;
@@ -4268,12 +4285,12 @@
   if (rc != SQLITE_OK)
   {
     const char* localError = sqlite3_errmsg((sqlite3*) m_db);
-    throw wxSQLite3Exception(rc, wxString::FromUTF8(localError));
+    WXSQLITE3_ERROR(rc, wxString::FromUTF8(localError), wxSQLite3StringCollection());
   }
   return wxSQLite3StringCollection(collectionName, p);
 #else
   wxUnusedVar(collectionName);
-  throw wxSQLite3Exception(WXSQLITE_ERROR, wxERRMSG_NOCOLLECTIONS);
+  WXSQLITE3_ERROR(WXSQLITE_ERROR, wxERRMSG_NOCOLLECTIONS, wxSQLite3StringCollection());
 #endif // WXSQLITE3_USE_NAMED_COLLECTIONS
 }
 
