Description: Fix build CREngine lvfntman
Author: Denis Kandrashin <mail@lintest.ru>
========================================================================================
Index: 3rdparty/crengine/include/lvfntman.h
===================================================================
--- 3rdparty/crengine/include/lvfntman.h.orig
+++ 3rdparty/crengine/include/lvfntman.h
@@ -266,7 +266,7 @@ public:
     /// get fallback font face (returns empty string if no fallback font is set)
     virtual lString8 GetFallbackFontFace() { return lString8::empty_str; }
     /// returns fallback font for specified size
-    virtual LVFontRef GetFallbackFont(int size) { }
+    virtual LVFontRef GetFallbackFont(int size) { return LVFontRef(); }
     /// registers font by name
     virtual bool RegisterFont( lString8 name ) = 0;
     /// initializes font manager
Index: 3rdparty/crengine/src/lvfntman.cpp
===================================================================
--- 3rdparty/crengine/src/lvfntman.cpp.orig
+++ 3rdparty/crengine/src/lvfntman.cpp
@@ -1499,6 +1499,7 @@ public:
             _fallbackFontFace = face;
             return !_fallbackFontFace.empty();
         }
+        return true;
     }
 
     /// get fallback font face (returns empty string if no fallback font is set)
