Description: Fix build against libpng 1.4.
Author: Kyrill Detinov <lazy.kent@opensuse.org>
Index: 3rdparty/crengine/src/lvimg.cpp
===================================================================
--- 3rdparty/crengine/src/lvimg.cpp.orig
+++ 3rdparty/crengine/src/lvimg.cpp
@@ -704,7 +704,7 @@ bool LVPngImageSource::Decode( LVImageDe
             png_set_palette_to_rgb(png_ptr);
 
         if (color_type == PNG_COLOR_TYPE_GRAY && bit_depth < 8)
-#if PNG_LIBPNG_VER_RELEASE==7
+#if (PNG_LIBPNG_VER_MAJOR == 1) && (PNG_LIBPNG_VER_MINOR < 4)
             png_set_gray_1_2_4_to_8(png_ptr);
 #else
             png_set_expand_gray_1_2_4_to_8(png_ptr);
